### DISCUSSION TITLE
TS: Narrowing a Union Type

### DISCUSSION THREAD

**Discussion by tobiasschweizer** [dotansimha/graphql-code-generator#10173] [Q&A] [answered]
Hi there,

I have a GraphQL query `fetch` that returns a union type `ResultUnion` which groups types such as `Book`, `ScholarlyArticle` etc.

```gql
query FETCH($iri: ID!) {
        
        fetch(iri: $iri) {
            __typename
        
            ... on ScholarlyArticle {
                iri
                name
                abstract
                description
                author {
                    ... on Person {
                        iri
                        name
                    }
                    ... on Organization {
                        iri
                        name
                    }
                }
            }
            
            ... on Book {
                name
                description
            }
      }
}
```

This means that the result of `fetch` is either a `Book` or a `ScholarlyArticle` etc.

```ts
export type FetchQuery = {
    __typename?: 'Query',
    fetch: { __typename: 'ArchiveComponent', name: string, description?: string | null } | {
        __typename: 'ArchiveOrganization',
        name: string,
        description?: string | null
    } | { __typename: 'Book', name: string, description?: string | null } | {
        __typename: 'Dataset',
        name: string,
        description?: string | null,
        author: Array<{ __typename?: 'Organization', iri: string, name: string } | {
            __typename?: 'Person',
            iri: string,
            name: string
        }>
    } | { __typename: 'Organization', name: string, description?: string | null } | {
        __typename: 'Person',
        name: string,
        description?: string | null
    } | { __typename: 'ResearchProject', name: string, description?: string | null } | {
        __typename: 'ScholarlyArticle',
        iri: string,
        name: string,
        abstract?: string | null,
        description?: string | null,
        author: Array<{ __typename?: 'Organization', iri: string, name: string } | {
            __typename?: 'Person',
            iri: string,
            name: string
        }>
    }
};
```

In TypeScript, the concrete type can be narrowed down:

```ts
    if (data?.fetch.__typename === 'ScholarlyArticle') {

        data.fetch.author <-- it is a ScholarlyArticle 

    }
``` 

Is there a way to construct a type from `FetchQuery` that is just 

```ts
{
        __typename: 'ScholarlyArticle',
        iri: string,
        name: string,
        abstract?: string | null,
        description?: string | null,
        author: Array<{ __typename?: 'Organization', iri: string, name: string } | {
            __typename?: 'Person',
            iri: string,
            name: string
        }>
    }
``` 

using an approach like mapped types, meaning that there is a mechanism that takes `FetchQuery` and produces a type for `ScholarlyArticle` given the information `fetch.__typename === 'ScholarlyArticle'`?

Thanks!

**Comment by tobiasschweizer** [comment, marked as answer]
see https://stackoverflow.com/questions/79112838/how-to-create-a-mapped-type-for-a-narrowed-union-type

**Reply by Katona** [reply]
The linked SO question no longer exists, is this accessible somewhere else?

**Reply by tobiasschweizer** [reply]
https://tsplay.dev/WvZbAW (I am not the author)

**Reply by Katona** [reply]
Thanks! It is still not clear how to use with query types with 'nested' unions, like yours for instance.

**Reply by tobiasschweizer** [reply]
This works for me:

You'd have to rename `typename` to `__typename` in the example code from the playground.

`type ResultUnion = FetchQuery['fetch'];`

![Screenshot 2025-06-06 at 14 41 33](https://github.com/user-attachments/assets/399682e0-ffde-49fa-8a47-827b33c3b913)

And then you can use this type, e.g.,
`const a: Book = { __typename: 'Book', name: 'An interesting book', description: 'Very interesting indeed'}`

`ResultUnion` is a union of all the types the query can possibly return and then the mapped type [Extract](https://www.typescriptlang.org/docs/handbook/utility-types.html#extracttype-union) is used and applies the `__typename` as the discriminator to extract the types you want.

You can also do this if you want: `type BookOrDataset = Specific<ResultUnion, 'Book' | 'Dataset'>`

**Reply by tobiasschweizer** [reply]
@Katona let me know if it works for you

**Reply by Katona** [reply]
This is clear, my question referred to the case where a union type is being in a nested field (or many nested fields). For instance, how would you access the field `fetch.author.name` in your initial example? Let's pretend there is a third type of the union of `author` where there is no `name` field so you _have to_ narrow.
In my case I have nested fields to which the 'path' leads through many fields having union types unfortunately.

**Reply by tobiasschweizer** [reply]
This would be a thing I don't have to address since the schema is consistent. Everything is derived from `schema:Thing` for which some basic props like `name` are defined so in case of an author, there would always be a `name`. There are specific props for a person like a `givenName` and to get that you would have to check at run time that it is a person. I think also if we had a parameter in the query that would exclude organizations as authors, the type would not reflect that.

Basically, I use these specific types in my React templates for the top level types like Book etc. so there is some basic type checking. 

Maybe you can provide your GraphQL-schema so I get a better understanding of your use case?

**Reply by tobiasschweizer** [reply]
```ts
export type PersonAuthor = Extract<Book['author'][number], { __typename: 'Person' }>
export type OrgAuthor = Extract<Book['author'][number], { __typename: 'Organization' }>
```

Is this helpful?
